dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Storages/DeltaMerge/BitmapFilter/BitmapFilter.h>
+#include <Storages/DeltaMerge/Segment.h>
+
+namespace DB::DM
+{
+ArrayBitmapFilter::ArrayBitmapFilter(UInt64 size_, SegmentSnapshotPtr snapshot_)
+    : filter(size_, 0)
+    , snap(snapshot_)
+{}
+
+void ArrayBitmapFilter::set(const ColumnPtr & col)
+{
+    for (size_t i = 0; i < col->size(); i++)
+    {
+        UInt64 row_id = col->getUInt(i);
```

Comment:
Please don't retain the `row_id` by calling a virtual function, i.e. `getUInt`. It is very slow. Check `toColumnVectorDataPtr`.